import os
from os.path import expanduser

from PyQt4.QtCore import QRegExp
from PyQt4.QtGui import QWidget, QRegExpValidator, QValidator

from gen.ui_new_project import UiNewProject
from misc.filedialog import FileDialog
from misc.messagebox import MessageBox


class NewProjectView(QWidget):
    def __init_(self, project_controller):
        super(NewProjectView, self).__init_()
        self.project_controller = project_controller
        self.ui = None
        self.validator = None
        self.build_ui()

    def build_ui(self):
        self.ui = UiNewProject()
        self.ui.setup_ui()
        self.ui.lineEdit_location.setText(expanduser("~"))
        regex = QRegExp("\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}")
        self.validator = QRegExpValidator(regex, self.ui.lineEdit_ip)
        self.ui.lineEdit_ip.setValidator(self.validator)
        self.ui.button_cancel.clicked.connect(self.clicked_cancel)
        self.ui.button_create.clicked.connect(self.clicked_create)
        self.ui.button_browse.clicked.connect(self.clicked_browse)

    def exec_(self):
        return self.ui.exec_()

    def clicked_cancel(self):
        self.ui.reject()

    def clicked_browse(self):
        self.ui.lineEdit_location.setText(FileDialog().get_existing_directory())

    def clicked_create(self):
        value = self.check_values()
        if value == 1:
            self.project_controller = self.project_controller.create_project(
                    self.ui.lineEdit_name.text(),
                    self.ui.lineEdit_ip.text(),
                    self.ui.lineEdit_location.text()
            )
            self.ui.done(1)
        elif value == -1:
            MessageBox().critical("Error", "You must to complete the information")
        elif value == -2:
            MessageBox().critical("Error", "You have to add an IP")
        elif value == -3:
            MessageBox().critical("Error", "Exist a project with same name on this path")

    def check_values(self):
        if self.ui.lineEdit_name.text():
            if self.check_ip():
                if self.check_location():
                    return 1
                else:
                    return -3
            else:
                return -2
        else:
            return -1

    def check_ip(self):
        # Verify if the IP "matches" the regular expression
        #TODO Falta comprobar que podemos hacer ping a esa IP, es decir que el host esta vivo
        state, pos = self.validator.validate(self.ui.lineEdit_ip.text(), 0)
        return state == QValidator.Acceptable

    def check_location(self):
        location = str(self.ui.lineEdit_location.text())
        name = str(self.ui.lineEdit_name.text())
        path = os.path.join(location, name + ".ini")
        return not os.path.exists(path) and os.access(str(self.ui.lineEdit_location.text()), os.W_OK)
